pkg/mutation/mutation_function.go

```diff
@@ -0,0 +1,140 @@
+	}
+	pathEntry := m.Path().Nodes[depth]
+	switch t := pathEntry.Type(); t {
+	case parser.ObjectNode:
+		next, ok := current.(map[string]interface{})[pathEntry.(*parser.Object).Reference]
+		if !ok {
+			next = createMissingElement(m, current, previous, depth)
+		}
+		if err := mutate(m, next, current, depth+1); err != nil {
+			return err
+		}
+		return nil
+	case parser.ListNode:
+		elementFound := false
+		listPathEntry := pathEntry.(*parser.List)
+		glob := listPathEntry.Glob
+		key := listPathEntry.KeyField
+		for _, listElement := range current.([]interface{}) {
+			if glob {
+				if err := mutate(m, listElement, current, depth+1); err != nil {
+					return err
+				}
+				elementFound = true
+			} else if elementValue, ok := listElement.(map[string]interface{})[key]; ok {
+				if *listPathEntry.KeyValue == elementValue {
+					if err := mutate(m, listElement, current, depth+1); err != nil {
+						return err
+					}
+					elementFound = true
+				}
+			}
+		}
+		// If no matching element in the array was found in non Globbed list, create a new element
+		if !listPathEntry.Glob && !elementFound {
+			next := createMissingElement(m, current, previous, depth)
+			if err := mutate(m, next, current, depth+1); err != nil {
+				return err
+			}
+		}
+	default:
+		return fmt.Errorf("Unrecognized type: %v", t)
+	}
+	return nil
+}
+
+func addValue(m Mutator, current interface{}, previous interface{}, depth int) error {
+	pathEntry := m.Path().Nodes[depth]
+	switch t := pathEntry.Type(); t {
+	case parser.ObjectNode:
+		key := pathEntry.(*parser.Object).Reference
+		switch m.(type) {
+		case *AssignMetadataMutator:
+
+			if elementValue, found, _ := unstructured.NestedString(current.(map[string]interface{}), key); found {
+				log.Info("Mutated value already present", "field", key, "value", elementValue)
+				return nil
+			}
+		}
+		value, err := m.Value()
+		if err != nil {
+			return err
+		}
+		if err = unstructured.SetNestedField(current.(map[string]interface{}), value, key); err != nil {
+			log.Error(err, "Failed to mutate object", "field", key, "value", value)
+		}
+
+	case parser.ListNode:
+		return setListElementToValue(m, current, previous, depth)
+	}
+	return nil
+}
+
+func setListElementToValue(m Mutator, current interface{}, previous interface{}, depth int) error {
+	listPathEntry := m.Path().Nodes[depth].(*parser.List)
+
+	if listPathEntry.Glob {
+		return fmt.Errorf("Last path entry can not be globbed")
+	}
+	key := listPathEntry.KeyField
+	keyValue := *listPathEntry.KeyValue
+	newValue, err := m.Value()
+	if err != nil {
+		log.Error(err, "Error getting mutator value")
+	}
+
+	elementFound := false
+	for _, listElement := range current.([]interface{}) {
+		if elementValue, found, _ := unstructured.NestedString(listElement.(map[string]interface{}), key); found && keyValue == elementValue {
+			elementFound = true
+			if err := unstructured.SetNestedField(listElement.(map[string]interface{}), newValue, key); err != nil {
```

Comment:
Here we can turn the above line

`for _, listElement := range currentList` (I'm assuming we do a safe type cast above)

into

`for i, listElement := range currentList`

then this line can be re-written as:

```
newValueMap, ok := newValue.(map[string]interface{})
if !ok {
   return errors.New("all list items must be objects")
}
newKeyVal, ok := newValue[key]
if !ok || keyVal != keyValue {
   return error
}
currentList[i] = newValue
```

The above logic only allows us to change the key value, which we definitely don't want to be able to do.